cmake/Gems.cmake

```diff
@@ -87,62 +87,37 @@ endmacro()
-function(resolve_all_dependent_gem_names_recursive gem_name input_resolved_gem_names output_resolved_gem_names)
+function(get_all_gem_dependencies gem_name output_resolved_gem_names)
 
-    # Stop if the gem dependency has already been resolved
-    list(FIND input_resolved_gem_names ${gem_name} found_gem_name_index)
-    if (found_gem_name_index GREATER_EQUAL 0)
-        # Recursive stop condition
+    get_property(gem_dependencies GLOBAL PROPERTY GEM_DEPENDENCIES_"${gem_name}")
+    if (gem_dependencies)
+        # The gem dependency for ${gem_name} has been calculated, return the cached property
+        set(${output_resolved_gem_names} ${gem_dependencies} PARENT_SCOPE)
         return()
-    endif()
-
-    # Attempt to read the dependencies for the gem
-    get_property(gem_path GLOBAL PROPERTY "@GEMROOT:${gem_name}@")
-    if (gem_path)
-        o3de_read_json_array(gem_dependencies "${gem_path}/gem.json" "dependencies")
     else()
-        return()
-    endif()
-
-    set(resolved_gem_names ${input_resolved_gem_names})
-
-    # Read in the dependencies recursively
-    foreach(enabled_gem_dependent_gem_name IN LISTS gem_dependencies)
-        list(APPEND resolved_gem_names "${enabled_gem_dependent_gem_name}")
-        resolve_all_dependent_gem_names_recursive(${enabled_gem_dependent_gem_name} ${resolved_gem_names} resolved_gem_names)
-    endforeach()
-
-    list(REMOVE_DUPLICATES resolved_gem_names)
-
-    set(${output_resolved_gem_names} ${resolved_gem_names} PARENT_SCOPE)
-
-endfunction()
+        # The gem dependency for ${gem_name} has not been calculated. 
+        # First read in the dependencies from the gem.json ifpossible
+        get_property(gem_path GLOBAL PROPERTY "@GEMROOT:${gem_name}@")
+        o3de_read_json_array(gem_dependencies "${gem_path}/gem.json" "dependencies")
 
-# resolve_all_dependent_gem_names
-#
-# Given a list of gem names, resolve all the dependent gems and build a list of additional
-# gems that need to be included for full dependency coverage
-#
-# \arg:input_gem_name_list(LIST) - The list of gem names to resolve all the dependencies
-# \arg:out_gem_name_list(LIST) - The list of additional gems that are not currently in the list to consider
-function(resolve_all_dependent_gem_names input_gem_name_list out_gem_name_list)
-
-    # Build up the list of all dependent gems (recursively)
-    unset(all_gems_name_list)
-    foreach(gem_name IN LISTS input_gem_name_list)
-        unset(dependent_gems_for_gem)
-        resolve_all_dependent_gem_names_recursive(${gem_name} "" dependent_gems_for_gem)
-        list(APPEND all_gems_name_list ${dependent_gems_for_gem})
-    endforeach()
+        # Set the property for the gem so that we won't encounter a possible infinite recursion
+        # when evaluating the gem dependencies' dependency
+        set_property(GLOBAL PROPERTY GEM_DEPENDENCIES_"${gem_name}" ${gem_dependencies})
```

Comment:
This seems kind of dangerous to partially set the gem name dependencies GLOBAL property part way through.
This would result in any gem name right before the start of a cycle receiving a partial list of the gem name dependencies, which could include itself.
ex. `Atom_RPI -> Atom_RHI -> Atom_RPI`.

Instead what could be done is to implement cycle detection logic by appending the visited gem names to a function scoped variable using `list(APPEND)`. 
If a cycle is detected, then a FATAL_ERROR can be logged as there would be no way to properly resolve the gem dependencies at that point.